### ISSUE TITLE
ObjectDisposedException (for FaultTolerantAmqpObject) in AmqpTransportHandler

### ISSUE THREAD

**Issue by varunpuranik** [Azure/azure-iot-sdk-csharp#785]
## Description of the issue:
We are getting a ObjectDisposedException for the FaultTolerantAmqpObject in the AmqpTransportHandler, typically when subscribing to Methods/Twins desired properties. It does not repro every time, but we are seeing it regularly.  

## Console log of the issue:
The call stack for the issue looks like this - 
 System.ObjectDisposedException : Cannot access a disposed object.
2019-01-30T23:33:14.0857865Z Object name: 'FaultTolerantAmqpObject`1'.
2019-01-30T23:33:14.0858589Z Stack Trace:
2019-01-30T23:33:14.0859193Z    at Microsoft.Azure.Devices.Client.Transport.AmqpTransportHandler.EnableTwinPatchAsync(CancellationToken cancellationToken)
2019-01-30T23:33:14.0859539Z    at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.<>c__DisplayClass22_0.<<ExecuteWithErrorHandlingAsync>b__0>d.MoveNext()
2019-01-30T23:33:14.0860066Z --- End of stack trace from previous location where exception was thrown ---
2019-01-30T23:33:14.0860319Z    at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.ExecuteWithErrorHandlingAsync[T](Func`1 asyncOperation)
2019-01-30T23:33:14.0860511Z    at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.<>c__DisplayClass26_0.<<EnableTwinPatchAsync>b__0>d.MoveNext()
2019-01-30T23:33:14.0861264Z --- End of stack trace from previous location where exception was thrown ---
2019-01-30T23:33:14.0861486Z    at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.EnableTwinPatchAsync(CancellationToken cancellationToken)
2019-01-30T23:33:14.0861713Z    at Microsoft.Azure.Devices.Client.InternalClient.SetDesiredPropertyUpdateCallbackAsync(DesiredPropertyUpdateCallback callback, Object userContext, CancellationToken cancellationToken)
2019-01-30T23:33:14.0861943Z    at Microsoft.Azure.Devices.Client.InternalClient.SetDesiredPropertyUpdateCallbackAsync(DesiredPropertyUpdateCallback callback, Object userContext)
2019-01-30T23:33:14.0862533Z    at Microsoft.Azure.Devices.Edge.Hub.CloudProxy.Test.CloudProxyT

**Comment by myagley** [comment]
Is there any update on this issue? We are seeing it quite a bit in the IoT Edge CI test runs.

**Comment by prmathur-microsoft** [comment]
We are WIP for this issue. Could you try out against private branch https://github.com/Azure/azure-iot-sdk-csharp/tree/amqp_redesign_preview and let us know if it fixes your issue? We plan to close this by EOM.

**Comment by perahlberg** [comment]
We have a similar problem, probably related to the same issue. 
We are using AMQP multiplexor. Set up as this:

var auth = new DeviceAuthenticationWithSharedAccessPolicyKey(webSocketSession.ApplianceId, WebSocketGatewaySettings.Instance.IotHubConnectionParts.SharedAccessKeyName, WebSocketGatewaySettings.Instance.IotHubConnectionParts.AccessPolicyKey);
var transportSettings = new AmqpTransportSettings(Microsoft.Azure.Devices.Client.TransportType.Amqp_Tcp_Only)
                {
                    AmqpConnectionPoolSettings = new AmqpConnectionPoolSettings()
                    {
                        Pooling = true,
                    }
                };

instance._client = DeviceClient.Create(WebSocketGatewaySettings.Instance.IotHubConnectionParts.Hostname, auth, new ITransportSettings[] { transportSettings });

instance._client.SetConnectionStatusChangesHandler(async (connectionStatus, connectionStatusChangeReason) =>
                {
// removed fro CLARITY
                });

              instance._client.SetDesiredPropertyUpdateCallbackAsync(webSocketSession.OnDesiredPropertyUpdateCallback, null).Wait();
instance._client.SetMethodDefaultHandlerAsync(webSocketSession.OnMethodDefaultHandler, null).Wait();

The problem begins when we random disconnect connected appliance and using the CloseAsync() method. This will (after a will) make all connected appliances fail with:
Cannot access a disposed object. Object name: 'FaultTolerantAmqpObject`1'.
When for example trying to send telemetry or sending state. Same error occur when trying to reconnect any appliance. 
In short the entirety multiplexor is unstable! 
This happen only is we using the CloseAsync() not if we only use Dispose().
We need to use the CloseAsync() for getting the correct “connected” state value fast in the IoTHub.
This is a major issue for Use.

I tried with the amqp_redesign_preview and the it seems to work!
Hop this helps!

**Comment by CIPop** [comment]
Fixed in v1.20.0

**Comment by az-iot-builder-01** [comment]
@prmathur-microsoft, @perahlberg, @CIPop, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
